index/postings.go

```diff
@@ -117,9 +122,61 @@ func (p *memPostings) ensureOrder() {
 	p.ordered = true
 }
 
-// add adds a document to the index. The caller has to ensure that no
-// term argument appears twice.
-func (p *memPostings) add(id uint64, lset labels.Labels) {
+// Delete removes all ids in the given map from the postings lists.
+func (p *MemPostings) Delete(deleted map[uint64]struct{}) {
+	var keys []labels.Label
+
+	p.mtx.RLock()
+	for l := range p.m {
+		keys = append(keys, l)
+	}
+	p.mtx.RUnlock()
+
+	for _, l := range keys {
```

Comment:
I am pretty sure these keys can go outdated pretty quick in race conditions, between this loop iterations and the initial keys append.

The only result would be new key in `p.m` being added in the meantime and not covered by this Delete method.. I guess this is fine?

Reply:
Anything added after we got the keys should by definition be something relevant to the new block range and thus not being relevant for GC.

Reply:
Can you add that as a comment? I did not understand why you were selecting the keys first until I read this.